File: bindings/c/src/label.cpp
// Copyright (c) 2025 Daniel T. McGinnis
// SPDX-License-Identifier: MIT

#include <nux/label.h>
#include "handles.hpp"

NUXC_EXTERNC
nux_label_t*
nux_label_create (
    nux_window_element_t* parent
) {
    auto element{nux_impl_map(parent)};

    if (
        auto enclosing_layout{dynamic_cast<nux::layout*>(element)};
        enclosing_layout
    ) {
        auto result{
            new nux_label_t{
                nux::label{*enclosing_layout},
                nux::string{}
            }
        };
        return result;
    }

    if (
        auto enclosing_group_box{dynamic_cast<nux::group_box*>(element)};
        enclosing_group_box
    ) {
        auto result{
            new nux_label_t{
                nux::label{
                    enclosing_group_box->grab_native_handle(),
                    enclosing_group_box->grab_content_layout()
                },
                nux::string{}
            }
        };
        return result;
    }

    return nullptr;
}

NUXC_EXTERNC
void
nux_label_destroy (
    nux_label_t* label
) {
    delete label;
}

NUXC_EXTERNC
nux_widget_t*
nux_label_upcast (
    nux_label_t* label
) {
    return nux_impl_map(static_cast<nux::widget*>(nux_impl_map(label)));
}

NUXC_EXTERNC
const nuxchar_t*
nux_label_get_text (
    nux_label_t* label,
    int* size
) {
    label->text = label->handle.get_text();
    if (size) {
        *size = label->text.size();
    }
    return reinterpret_cast<const nuxchar_t*>(label->text.data());
}

NUXC_EXTERNC
void
nux_label_set_text (
    nux_label_t* label,
    const nuxchar_t* new_text
) {
    label->handle.set_text(
        reinterpret_cast<nux::string_view::const_pointer>(new_text)
    );
}

File: bindings/rs/nux/src/label.rs
// Copyright (c) 2025 Daniel T. McGinnis
// SPDX-License-Identifier: MIT

use std::ffi::*;
use std::rc::*;
use std::cell::*;
use crate::bindings::*;
use crate::window_element::*;
use crate::widget::*;

#[derive(Clone)]
pub struct Label {
    widget_handle_ref: WidgetHandleReference
}

impl Label {
    pub fn get_text (
        &self
    ) -> String {
        unsafe {
            let mut size: std::os::raw::c_int = 0;
            let data: *const nuxchar_t = nux_label_get_text(
                self.widget_handle_ref.borrow().handle.native_handle
                    as *mut nux_label_t,
                &mut size as *mut std::os::raw::c_int
            );
            return str::from_utf8(
                std::slice::from_raw_parts(
                    data as *const u8,
                    size as usize
                )
            ).unwrap().to_string();
        }
    }

    pub fn set_text (
        &self,
        text: &str
    ) {
        unsafe {
            nux_label_set_text(
                self.widget_handle_ref.borrow().handle.native_handle
                    as *mut nux_label_t,
                CString::new(text).unwrap().as_ptr()
                    as *const u8
            );
        };
    }
}

impl WindowElement for Label {
    fn grab_handle (
        &self
    ) -> WindowElementHandle {
        return self.grab_widget_handle().borrow().handle;
    }
}

impl Widget for Label {
    fn grab_widget_handle (
        &self
    ) -> WidgetHandleReference {
        return self.widget_handle_ref.clone();
    }
}

impl SimpleWidget for Label {
    fn new (
        parent: &WindowElementHandle
    ) -> Self {
        return Self {
            widget_handle_ref: Rc::new(
                RefCell::new(
                    WidgetHandle {
                        handle: WindowElementHandle {
                            native_handle: unsafe {
                                nux_label_create(
                                    parent.native_handle
                                ) as *mut nux_window_element_t
                            }
                        },
                        kind: NativeWidgetKind::Label,
                        graphics_box_render_event_handler: None,
                        push_button_trigger_event_handler: None,
                        check_box_toggle_event_handler: None,
                        radio_button_toggle_event_handler: None
                    }
                )
            )
        };
    }
}